apps/astarte_data_updater_plant/lib/astarte_data_updater_plant/amqp_events_producer.ex

```diff
@@ -44,73 +41,84 @@ defmodule Astarte.DataUpdaterPlant.AMQPEventsProducer do
     GenServer.call(__MODULE__, {:declare_exchange, exchange}, 60_000)
   end
 
-  # Server callbacks
+  # Gets a connection worker out of the connection pool, if there is one available
+  # takes a channel out of it channel pool, if there is one available
+  defp do_connect(state) do
+    conn = ExRabbitPool.get_connection_worker(:events_producer_pool)
+
+    case ExRabbitPool.checkout_channel(conn) do
+      {:ok, channel} ->
+        try_to_connect(channel, state)
 
-  def init(_args) do
-    rabbitmq_connect(false)
+      {:error, _reason} ->
+        schedule_connect()
+        {:noreply, state}
+    end
   end
 
-  def terminate(_reason, %Channel{conn: conn} = chan) do
-    Channel.close(chan)
-    Connection.close(conn)
+  # When successfully checks out a channel, sets up exchange, and monitors it to handle crashes and reconnections
+  defp try_to_connect(channel, _state) do
+    %{pid: channel_pid} = channel
+
+    case @adapter.declare_exchange(channel, Config.events_exchange_name!(),
+           type: :direct,
+           durable: true
+         ) do
+      :ok ->
+        Process.monitor(channel_pid)
+
+        {:noreply, channel}
+
+      {:error, reason} ->
+        Logger.warn("RabbitMQ Connection error: #{inspect(reason)}",
+          tag: "events_producer_conn_err"
+        )
+
+        schedule_connect()
```

Comment:
Here you're returning the `{:noreply, :not_connected}` returned by `schedule_connect()`, at line 54 you're ignoring it and returning another state instead.

`schedule_connect()` should used only for side effects (i.e. just sending the message) and the state must be handled outside of it (this also makes it clearer since you don't have to jump to the implementation of `schedule_connect` to see what it will return)